src/core/include/openvino/core/model.hpp

```diff
@@ -31,15 +31,16 @@ std::shared_ptr<Model> clone_model(const Model& func, std::unordered_map<Node*,
 
 namespace frontend {
 class FrontEnd;
-}
+class FrontEndSharedData;
+}  // namespace frontend
 
 class ModelAccessor;
 /// A user-defined model.
 class OPENVINO_API Model : public std::enable_shared_from_this<Model> {
     friend class frontend::FrontEnd;
     friend OPENVINO_API std::shared_ptr<Model> clone_model(const Model& func,
                                                            std::unordered_map<Node*, std::shared_ptr<Node>>& node_map);
-    std::shared_ptr<void> m_shared_object;  // Frontend plugin shared object handle.
+    std::shared_ptr<frontend::FrontEndSharedData> m_shared_data;  // Frontend plugin shared object handle.
```

Comment:
Does `shared_ptr<void>` not work here?
@ilyachur Previously, `shared_ptr<void>` could hide everything and user doesn't even see that these is a `frontend::FrontEndSharedData` class declaration. Now public header shows that there is some `FrontEndSharedData` class, which doesn't look well

Reply:
@mbencer Why cannot we use the previous approach with `std::shared_ptr<void>` to hold the Frontend?
But at the same moment, `Frontend` object should hold all extensions.
I think it should work.

Reply:
@mbencer Could you try to change the approach?

Reply:
Yes something like this.
Why doesn't it solve the issue?
From my point of view we don't have a big difference in these approaches:

- Current approach: `ov::Model` holds frontend and extension objects.
- In my approach: `ov::Model` holds frontend, frontend holds extensions.

These approaches should work in the same way, if first of all we remove nodes and other model objects and only after that remove frontend.

Reply:
In current approach the order of dtor is as following:
```
- Model dtor
- FrontEnd::m_actual dtor
- FrontEnd dtor
- Model copy dtor
- so extension dtor, dll is closed
```
Based on your proposition I'am already having:
```
- Model dtor
- so extension dtor, dll is closed
- Model copy dtor
- segfault is thrown
```
I'm trying to check why

Reply:
@ilyachur I think that your appoarch is correct but only if we gurantee that `Frontend` is released after `Model`. We can achieve it with approach like https://github.com/mbencer/openvino/commit/19b36365997fd3b111346b901c5fc6b0ae634a2f, but at least locally it makes some other python tests broken.
Please notice that in shared object passed from `Frontend` to `Model` we are passing now only a pointer to the dynamic library with Frontend (`get_so_pointer` - https://github.com/openvinotoolkit/openvino/blob/93722fe101700218c3b0ca126602d90128ccced4/src/frontends/common/src/manager.cpp#L48), but `Frontend` object can be released ealier (before `Model`).

Should we go to the direction like from https://github.com/mbencer/openvino/commit/19b36365997fd3b111346b901c5fc6b0ae634a2f (for me not very clear) or keep the current solution (with foward declaration or void approach?)

Reply:
@ilyachur The other approach can be like from [933fcf9](https://github.com/openvinotoolkit/openvino/pull/10749/commits/933fcf9e062d1164a53975453ead9bb3f44623c7)

Reply:
The last approach looks ok for me because we don't introduce any new additional structures